src/Orc.NuGetExplorer/Orc.NuGetExplorer.Shared/Services/ApiPackageRegistry.cs

```diff
@@ -0,0 +1,88 @@
+﻿// --------------------------------------------------------------------------------------------------------------------
+// <copyright file="ApiPackageRegistry.cs" company="WildGums">
+//   Copyright (c) 2008 - 2017 WildGums. All rights reserved.
+// </copyright>
+// --------------------------------------------------------------------------------------------------------------------
+
+namespace Orc.NuGetExplorer
+{
+    using System;
+    using System.Collections.Generic;
+    using Catel;
+    using Catel.Logging;
+    using NuGet;
+
+    internal sealed class ApiPackageRegistry : IApiPackageRegistry
+    {
+        #region Fields
+        private static readonly ILog Log = LogManager.GetCurrentClassLogger();
+
+        private readonly Dictionary<string, SemanticVersion> _apiPackages = new Dictionary<string, SemanticVersion>();
+
+        private readonly object _syncObj = new object();
+        #endregion
+
+        #region Methods
+        public void Register(string packageName, string version)
+        {
+            Argument.IsNotNullOrWhitespace(() => packageName);
+            var semanticVersion = SemanticVersion.Parse(version);
+
+            lock (_syncObj)
+            {
+                if (_apiPackages.TryGetValue(packageName, out var storedSemanticVersion))
+                {
+                    throw Log.ErrorAndCreateException<ArgumentException>("The api package '{0}' is already registered with version '{1}'", packageName, storedSemanticVersion);
+                }
+
+                _apiPackages.Add(packageName, semanticVersion);
+            }
+        }
+
+        public bool IsRegistered(string packageName)
+        {
+            lock (_syncObj)
+            {
+                return _apiPackages.TryGetValue(packageName, out var _);
+            }
+        }
+
+        public void Validate(IPackageDetails package)
+        {
+            package.ApiValidations.Clear();
+            var innerPackage = ((PackageDetails) package).Package;
+            foreach (var dependencySet in innerPackage.DependencySets)
+            {
+                foreach (var dependency in dependencySet.Dependencies)
+                {
+                    lock (_syncObj)
```

Comment:
Isn't locking all the time more expensive than adding the lock outside the foreach?